Core/include/Acts/Utilities/GridBinFinder.hpp

```diff
@@ -22,30 +24,33 @@ namespace Acts {
 /// used to find both bins that could be bottom bins as well as bins that could
 /// be top bins, which are assumed to be the same bins. Does not take
 /// interaction region into account to limit z-bins.
-template <typename external_spacepoint_t>
-class BinFinder {
+template <std::size_t DIM>
+class GridBinFinder {
  public:
-  /// constructor
-  BinFinder() = delete;
-
-  BinFinder(const std::vector<std::pair<int, int>>& zBinNeighbors,
-            int numPhiNeighbors);
+  template <typename... args>
+  GridBinFinder(args&&... vals);
 
   /// Return all bins that could contain space points that can be used with the
   /// space points in the bin with the provided indices to create seeds.
   /// @param phiBin phi index of bin with middle space points
   /// @param zBin z index of bin with middle space points
   /// @param binnedSP phi-z grid containing all bins
-  boost::container::small_vector<std::size_t, 9> findBins(
-      std::size_t phiBin, std::size_t zBin,
-      const SpacePointGrid<external_spacepoint_t>* binnedSP) const;
+  template <typename stored_t, class... Axes>
+  boost::container::small_vector<std::size_t, Acts::detail::ipow(3, DIM)> findBins(
+      const std::array<std::size_t, DIM>& locPosition,
+      const Acts::Grid<stored_t, Axes...>& grid) const;
 
  private:
-  // This vector is provided by the user and is supposed to be a constant for
-  // all events. No point in making a copy
-  Acts::detail::RefHolder<const std::vector<std::pair<int, int>>>
-      m_zBinNeighbors;
-  int m_numPhiNeighbors = 1;
+  template <typename first_value_t, typename... vals>
+  void storeValue(first_value_t&& fv, vals&&... others);
+
+  std::array<std::pair<int, int>, DIM> getSizePerAxis(
+      const std::array<std::size_t, DIM>& locPosition) const;
+
+ private:
+  using stored_values_t = std::variant<int, std::vector<std::pair<int, int>>>;
```

Comment:
Keep in mind that the variant has the memory footprint of the vector even in case an integer is stored.

Reply:
good point